language/move-vm/runtime/src/data_operations.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use libra_logger::prelude::*;
+use libra_types::{
+    access_path::AccessPath,
+    vm_error::{sub_status, StatusCode},
+};
+use move_core_types::gas_schedule::{AbstractMemorySize, GasAlgebra, GasCarrier};
+use move_vm_types::{
+    data_store::DataStore,
+    loaded_data::types::FatStructType,
+    values::{GlobalValue, Struct, Value},
+};
+use vm::errors::{vm_error, Location, VMResult};
+
+//
+// Provides an implementation for data store bytecodes and guarantees proper invariants
+// for each operation. It uses the `DataStore` for the data operation but it makes sure
+// all operations are consistent to Move semantic.
```

Comment:
Just out of curiosity why do we want to have some logics to be in DataStore and the rest over here? Can we implement those logics directly in `data_cache.rs`?

Reply:
the idea here is that those operations are "vital" to the system and should not and are not provided by the cache and particularly by the trait (given the trait is not providing an implementation).
So basically the functions here are internal to the VM and are called by the bytecodes or the natives and give the guarantees of proper implementation over the `DataStore` or data caches.
Caches and in general implementations of `DataStore` offer a much more direct and simple managing of the cache deprived of semantic that is a VM concern